boot/reboot.go

```diff
@@ -0,0 +1,101 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"fmt"
+	"os/exec"
+	"time"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/osutil"
+)
+
+func getRebootArgsPathRuntime() string {
+	return "/run/systemd/reboot-param"
+}
+
+// GetRebootArgsPath is used so we can mock the path easily in tests
+var GetRebootArgsPath = getRebootArgsPathRuntime
+
+type RebootAction int
+
+func (a RebootAction) String() string {
+	switch a {
+	case RebootReboot:
+		return "system reboot"
+	case RebootHalt:
+		return "system halt"
+	case RebootPoweroff:
+		return "system poweroff"
+	default:
+		panic(fmt.Sprintf("unknown reboot action %d", a))
+	}
+}
+
+const (
+	RebootReboot RebootAction = iota
+	RebootHalt
+	RebootPoweroff
+)
+
+var (
+	shutdownMsg = i18n.G("reboot scheduled to update the system")
+	haltMsg     = i18n.G("system halt scheduled")
+	poweroffMsg = i18n.G("system poweroff scheduled")
+)
+
+func Reboot(action RebootAction, rebootDelay time.Duration, rebootInfo *RebootInfo) error {
```

Comment:
does this need some check to make sure it is mocked in tests?  maybe its usage is very specific and we don't need that

Reply:
I see your point here, but I am unsure on what safeguards to add. We could call `osutil.IsTestBinary()` and do not call `shutdown` if we are testing, but there are some tests that mock the call to the `shutdown` binary and then check if has been called, which I think is something we want to keep. The way it could work is by checking if the command somehow is being mocked, but not sure if there is something like that atm.

Reply:
I've done this now